Unable to upload binary file

Hello,

Thanks for the project - just been trying to use it to upload files but seeing something weird:

Create a file (this seems to be the easiest way to replicate the issue), naturally my actual file isn't random:
```sh
head -c 200000 < /dev/random > /tmp/foo.txt
```

Attempt to upload the file (using python 3.7, boto3==1.9.130)
```python
import boto3
client = boto3.client("s3", "http://<my_endpoint>:9090")
client.upload_file("/tmp/foo.txt", "<bucket_name>", "foo.txt")
```

The client raises a botocore.exceptions.ClientException:
```
botocore.exceptions.ClientError: An error occurred (500) when calling the PutObject operation (reached max retries: 4): Internal Server Error
```

Tailing the logs for the S3 container sheds a bit more light:

```
s3_1        | org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.NumberFormatException: For input string: "??8?oz??@)???]'c???M?'?"
s3_1        |   at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1013) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:919) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at javax.servlet.http.HttpServlet.service(HttpServlet.java:668) ~[javax.servlet-api-4.0.1.jar!/:4.0.1]
s3_1        |   at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:882) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at javax.servlet.http.HttpServlet.service(HttpServlet.java:750) ~[javax.servlet-api-4.0.1.jar!/:4.0.1]
s3_1        |   at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:873) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1623) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at com.adobe.testing.s3mock.KmsValidationFilter.doFilterInternal(KmsValidationFilter.java:86) ~[classes!/:na]
s3_1        |   at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:101) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:540) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) ~[jetty-security-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1701) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480) ~[jetty-servlet-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1668) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.Server.handle(Server.java:502) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267) ~[jetty-server-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305) ~[jetty-io-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) ~[jetty-io-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) ~[jetty-io-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) ~[jetty-util-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) ~[jetty-util-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) ~[jetty-util-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126) ~[jetty-util-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366) ~[jetty-util-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765) ~[jetty-util-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683) ~[jetty-util-9.4.15.v20190215.jar!/:9.4.15.v20190215]
s3_1        |   at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
s3_1        | Caused by: java.lang.NumberFormatException: For input string: "??8?oz??@)???]'c???M?'?"
s3_1        |   at java.base/java.lang.NumberFormatException.forInputString(Unknown Source) ~[na:na]
s3_1        |   at java.base/java.lang.Integer.parseInt(Unknown Source) ~[na:na]
s3_1        |   at com.adobe.testing.s3mock.util.AwsChunkDecodingInputStream.read(AwsChunkDecodingInputStream.java:79) ~[classes!/:na]
s3_1        |   at java.base/java.io.InputStream.read(Unknown Source) ~[na:na]
s3_1        |   at java.base/java.io.InputStream.read(Unknown Source) ~[na:na]
s3_1        |   at com.adobe.testing.s3mock.domain.FileStore.inputStreamToFile(FileStore.java:437) ~[classes!/:na]
s3_1        |   at com.adobe.testing.s3mock.domain.FileStore.putS3Object(FileStore.java:248) ~[classes!/:na]
s3_1        |   at com.adobe.testing.s3mock.FileStoreController.putObject(FileStoreController.java:476) ~[classes!/:na]
s3_1        |   at jdk.internal.reflect.GeneratedMethodAccessor26.invoke(Unknown Source) ~[na:na]
s3_1        |   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
s3_1        |   at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
s3_1        |   at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:189) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138) ~[spring-web-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:102) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:892) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:797) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1038) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1005) ~[spring-webmvc-5.1.6.RELEASE.jar!/:5.1.6.RELEASE]
s3_1        |   ... 49 common frames omitted
```

Looks like the upload handler is assuming that it's going to be given a multipart upload where the content-length is given before a semi-colon (in hex format?) but we're not passing a hex string there - instead it looks to be something like `??8?oz??@)???]'c???M?'?` (https://github.com/adobe/S3Mock/blob/master/server/src/main/java/com/adobe/testing/s3mock/util/AwsChunkDecodingInputStream.java#L79).

I wondered if this was a boto3 issue / my error but the exact same code works against the live S3 API (just removing the `endpoint_url` when creating the client). Any ideas?

--- a/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/FileStoreController.java
@@ -119,7 +119,8 @@ class FileStoreController {
 
   private static final String RANGES_BYTES = "bytes";
 
-  private static final String UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD";
+  private static final String STREAMING_AWS_4_HMAC_SHA_256_PAYLOAD =
+      "STREAMING-AWS4-HMAC-SHA256-PAYLOAD";
 
   private static final String RESPONSE_HEADER_CONTENT_TYPE = "response-content-type";
   private static final String RESPONSE_HEADER_CONTENT_LANGUAGE = "response-content-language";
@@ -140,7 +141,7 @@ class FileStoreController {
 
   private static final Comparator<String> KEY_COMPARATOR = Comparator.naturalOrder();
   private static final Comparator<BucketContents> BUCKET_CONTENTS_COMPARATOR =
-          Comparator.comparing(BucketContents::getKey, KEY_COMPARATOR);
+      Comparator.comparing(BucketContents::getKey, KEY_COMPARATOR);
 
   @Autowired
   private FileStore fileStore;
@@ -261,7 +262,8 @@ class FileStoreController {
   }
 
   /**
-   * Retrieve list of objects of a bucket see http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketGET.html
+   * Retrieve list of objects of a bucket see http://docs.aws.amazon
+   * .com/AmazonS3/latest/API/RESTBucketGET.html
    *
    * @param bucketName {@link String} set bucket name
    * @param prefix {@link String} find object names they starts with prefix
@@ -282,19 +284,19 @@ class FileStoreController {
       @RequestParam(required = false) final String delimiter,
       @RequestParam(name = "encoding-type", required = false) final String encodingtype,
       @RequestParam(name = "max-keys", defaultValue = "1000",
-              required = false) final Integer maxKeys,
+          required = false) final Integer maxKeys,
       final HttpServletResponse response) throws IOException {
     verifyBucketExistence(bucketName);
     if (maxKeys < 0) {
       throw new S3Exception(HttpStatus.BAD_REQUEST.value(), "InvalidRequest",
-              "maxKeys should be non-negative");
+          "maxKeys should be non-negative");
     }
     if (!StringUtils.isEmpty(encodingtype) && !"url".equals(encodingtype)) {
       throw new S3Exception(HttpStatus.BAD_REQUEST.value(), "InvalidRequest",
           "encodingtype can only be none or 'url'");
     }
 
-    boolean useUrlEncoding = Objects.equals("url", encodingtype);
+    final boolean useUrlEncoding = Objects.equals("url", encodingtype);
 
     try {
       List<BucketContents> contents = getBucketContents(bucketName, prefix);
@@ -321,10 +323,10 @@ class FileStoreController {
     return null;
   }
 
-  private List<BucketContents> applyUrlEncoding(List<BucketContents> contents) {
+  private List<BucketContents> applyUrlEncoding(final List<BucketContents> contents) {
     return contents.stream().map(c -> new BucketContents(UrlEncoded.encodeString(c.getKey()),
         c.getLastModified(), c.getEtag(), c.getSize(), c.getStorageClass(), c.getOwner())).collect(
-            Collectors.toList());
+        Collectors.toList());
   }
 
   /**
@@ -391,7 +393,7 @@ class FileStoreController {
           "encodingtype can only be none or 'url'");
     }
 
-    boolean useUrlEncoding = Objects.equals("url", encodingtype);
+    final boolean useUrlEncoding = Objects.equals("url", encodingtype);
 
     verifyBucketExistence(bucketName);
     try {
@@ -407,7 +409,8 @@ class FileStoreController {
         and then Amazon S3 ignores this parameter.
        */
       if (continuationToken != null) {
-        String continueAfter = fileStorePagingStateCache.get(continuationToken).get().toString();
+        final String continueAfter =
+            fileStorePagingStateCache.get(continuationToken).get().toString();
         filteredContents = getFilteredBucketContents(contents, continueAfter);
         fileStorePagingStateCache.evict(continuationToken);
       } else {
@@ -420,7 +423,7 @@ class FileStoreController {
         nextContinuationToken = UUID.randomUUID().toString();
         filteredContents = filteredContents.subList(0, maxKeys);
         fileStorePagingStateCache.put(nextContinuationToken,
-                                      filteredContents.get(maxKeys - 1).getKey());
+            filteredContents.get(maxKeys - 1).getKey());
       }
 
       final Set<String> commonPrefixes = new HashSet<>();
@@ -448,9 +451,9 @@ class FileStoreController {
       final String startAfter) {
     if (startAfter != null && !"".equals(startAfter)) {
       return contents
-              .stream()
-              .filter(p -> KEY_COMPARATOR.compare(p.getKey(), startAfter) > 0)
-              .collect(Collectors.toList());
+          .stream()
+          .filter(p -> KEY_COMPARATOR.compare(p.getKey(), startAfter) > 0)
+          .collect(Collectors.toList());
     } else {
       return contents;
     }
@@ -458,7 +461,7 @@ class FileStoreController {
 
   private List<BucketContents> getBucketContents(final String bucketName,
       final String prefix) throws IOException {
-    String encodedPrefix = null != prefix ? objectNameToFileName(prefix) : null;
+    final String encodedPrefix = null != prefix ? objectNameToFileName(prefix) : null;
 
     final List<S3Object> s3Objects = fileStore.getS3Objects(bucketName, encodedPrefix);
 
@@ -495,7 +498,7 @@ class FileStoreController {
           request.getContentType(),
           request.getHeader(HttpHeaders.CONTENT_ENCODING),
           inputStream,
-          isV4SigningEnabled(request),
+          isV4ChunkedWithSigningEnabled(request),
           userMetadata);
 
       addTagsFromReq(request, bucketName, filename);
@@ -512,13 +515,13 @@ class FileStoreController {
   }
 
   private void addTagsFromReq(final HttpServletRequest request,
-                              final String bucketName,
-                              final String filename) throws IOException {
+      final String bucketName,
+      final String filename) throws IOException {
     final String header = request.getHeader(HEADER_X_AMZ_TAGGING);
     if (header != null && !header.isEmpty()) {
       final List<Tag> tags = new ArrayList<>();
       new UrlEncoded(header)
-              .forEach((tag, values) -> tags.add(new Tag(tag, values.get(0))));
+          .forEach((tag, values) -> tags.add(new Tag(tag, values.get(0))));
 
       fileStore.setObjectTags(bucketName, filename, tags);
     }
@@ -533,9 +536,9 @@ class FileStoreController {
         ));
   }
 
-  private boolean isV4SigningEnabled(final HttpServletRequest request) {
+  private boolean isV4ChunkedWithSigningEnabled(final HttpServletRequest request) {
     final String sha256Header = request.getHeader(HEADER_X_AMZ_CONTENT_SHA256);
-    return sha256Header != null && !sha256Header.equals(UNSIGNED_PAYLOAD);
+    return sha256Header != null && sha256Header.equals(STREAMING_AWS_4_HMAC_SHA_256_PAYLOAD);
   }
 
   /**
@@ -575,7 +578,7 @@ class FileStoreController {
               filename,
               request.getContentType(),
               inputStream,
-              isV4SigningEnabled(request),
+              isV4ChunkedWithSigningEnabled(request),
               userMetadata,
               encryption,
               kmsKeyId);
@@ -743,7 +746,7 @@ class FileStoreController {
       final String key = URLDecoder.decode(substringBefore(param, "="), UTF_8.name());
       final String value = URLDecoder.decode(substringAfter(param, "="), UTF_8.name());
       return new SimpleImmutableEntry<>(key, value);
-    } catch (UnsupportedEncodingException e) {
+    } catch (final UnsupportedEncodingException e) {
       throw new AssertionError(UTF_8.name() + " is unknown");
     }
   }
@@ -959,8 +962,8 @@ class FileStoreController {
 
     final String uploadId = UUID.randomUUID().toString();
     fileStore.prepareMultipartUpload(bucketName, filename, request.getContentType(),
-            request.getHeader(HttpHeaders.CONTENT_ENCODING), uploadId,
-            TEST_OWNER, TEST_OWNER, userMetadata);
+        request.getHeader(HttpHeaders.CONTENT_ENCODING), uploadId,
+        TEST_OWNER, TEST_OWNER, userMetadata);
 
     return new InitiateMultipartUploadResult(bucketName, fileNameToObjectName(filename), uploadId);
   }
@@ -988,9 +991,9 @@ class FileStoreController {
 
     final List<MultipartUpload> multipartUploads =
         fileStore.listMultipartUploads().stream().map(m ->
-          new MultipartUpload(fileNameToObjectName(m.getKey()), m.getUploadId(), m.getOwner(),
-              m.getInitiator(), m.getInitiated()))
-        .collect(Collectors.toList());
+            new MultipartUpload(fileNameToObjectName(m.getKey()), m.getUploadId(), m.getOwner(),
+                m.getInitiator(), m.getInitiated()))
+            .collect(Collectors.toList());
 
     // the result contains all uploads, use some common value as default
     final int maxUploads = Math.max(1000, multipartUploads.size());
@@ -1099,7 +1102,7 @@ class FileStoreController {
         uploadId,
         partNumber,
         request.getInputStream(),
-        isV4SigningEnabled(request));
+        isV4ChunkedWithSigningEnabled(request));
 
     final HttpHeaders responseHeaders = new HttpHeaders();
     final String quotedEtag = "\"" + etag + "\"";
@@ -1357,13 +1360,13 @@ class FileStoreController {
    * @return encoded key
    */
   // @VisibleForTesting
-  static String objectNameToFileName(String objectName) {
-    char[] chars = objectName.toCharArray();
+  static String objectNameToFileName(final String objectName) {
+    final char[] chars = objectName.toCharArray();
 
-    int len = chars.length;
+    final int len = chars.length;
     StringBuffer buffer = null;
     for (int i = 0; i < len; i++) {
-      char c = chars[i];
+      final char c = chars[i];
 
       // the following characters need escaping
       if (c < ' ' || c >= 0x7f || c == '<' || c == '>' || c == ':' || c == '"' || c == '\\'
@@ -1389,12 +1392,12 @@ class FileStoreController {
   }
 
   // @VisibleForTesting
-  static String fileNameToObjectName(String encoded) {
+  static String fileNameToObjectName(final String encoded) {
     StringBuffer buffer = null;
 
-    char[] chars = encoded.toCharArray();
+    final char[] chars = encoded.toCharArray();
     for (int i = 0; i < chars.length; i++) {
-      char c = chars[i];
+      final char c = chars[i];
 
       if (c == '%') {
         if (buffer == null) {

--- a/server/src/main/java/com/adobe/testing/s3mock/domain/FileStore.java
+++ b/server/src/main/java/com/adobe/testing/s3mock/domain/FileStore.java
@@ -201,7 +201,7 @@ public class FileStore {
    * @param contentType The files Content Type.
    * @param contentEncoding The files Content Encoding.
    * @param dataStream The File as InputStream.
-   * @param useV4Signing If {@code true}, V4-style signing is enabled.
+   * @param useV4ChunkedWithSigningFormat If {@code true}, V4-style signing is enabled.
    *
    * @return {@link S3Object}.
    *
@@ -212,9 +212,9 @@ public class FileStore {
       final String contentType,
       final String contentEncoding,
       final InputStream dataStream,
-      final boolean useV4Signing) throws IOException {
-    return putS3Object(bucketName, fileName, contentType, contentEncoding, dataStream, useV4Signing,
-        Collections.emptyMap());
+      final boolean useV4ChunkedWithSigningFormat) throws IOException {
+    return putS3Object(bucketName, fileName, contentType, contentEncoding, dataStream,
+        useV4ChunkedWithSigningFormat, Collections.emptyMap());
   }
 
   /**
@@ -225,7 +225,7 @@ public class FileStore {
    * @param contentType The files Content Type.
    * @param contentEncoding The files Content Encoding.
    * @param dataStream The File as InputStream.
-   * @param useV4Signing If {@code true}, V4-style signing is enabled.
+   * @param useV4ChunkedWithSigningFormat If {@code true}, V4-style signing is enabled.
    * @param userMetadata User metadata to store for this object, will be available for the
    *     object with the key prefixed with "x-amz-meta-".
    *
@@ -238,7 +238,7 @@ public class FileStore {
       final String contentType,
       final String contentEncoding,
       final InputStream dataStream,
-      final boolean useV4Signing,
+      final boolean useV4ChunkedWithSigningFormat,
       final Map<String, String> userMetadata) throws IOException {
     final S3Object s3Object = new S3Object();
     s3Object.setName(fileName);
@@ -251,7 +251,7 @@ public class FileStore {
     final File objectRootFolder = createObjectRootFolder(theBucket, s3Object.getName());
 
     final File dataFile =
-        inputStreamToFile(wrapStream(dataStream, useV4Signing),
+        inputStreamToFile(wrapStream(dataStream, useV4ChunkedWithSigningFormat),
             objectRootFolder.toPath().resolve(DATA_FILE));
     s3Object.setDataFile(dataFile);
     s3Object.setSize(Long.toString(dataFile.length()));
@@ -271,9 +271,10 @@ public class FileStore {
     return s3Object;
   }
 
-  private InputStream wrapStream(final InputStream dataStream, final boolean useV4Signing) {
+  private InputStream wrapStream(final InputStream dataStream,
+      final boolean useV4ChunkedWithSigningFormat) {
     final InputStream inStream;
-    if (useV4Signing) {
+    if (useV4ChunkedWithSigningFormat) {
       inStream = new AwsChunkDecodingInputStream(dataStream);
     } else {
       inStream = dataStream;
@@ -314,7 +315,7 @@ public class FileStore {
    * @param fileName name of the File to be stored.
    * @param contentType The files Content Type.
    * @param dataStream The File as InputStream.
-   * @param useV4Signing If {@code true}, V4-style signing is enabled.
+   * @param useV4ChunkedWithSigningFormat If {@code true}, V4-style signing is enabled.
    * @param userMetadata User metadata to store for this object, will be available for the
    *     object with the key prefixed with "x-amz-meta-".
    * @param encryption The Encryption Type.
@@ -328,7 +329,7 @@ public class FileStore {
       final String fileName,
       final String contentType,
       final InputStream dataStream,
-      final boolean useV4Signing,
+      final boolean useV4ChunkedWithSigningFormat,
       final Map<String, String> userMetadata,
       final String encryption, final String kmsKeyId) throws IOException {
     final S3Object s3Object = new S3Object();
@@ -344,7 +345,7 @@ public class FileStore {
     final File objectRootFolder = createObjectRootFolder(theBucket, s3Object.getName());
 
     final File dataFile =
-        inputStreamToFile(wrapStream(dataStream, useV4Signing),
+        inputStreamToFile(wrapStream(dataStream, useV4ChunkedWithSigningFormat),
             objectRootFolder.toPath().resolve(DATA_FILE));
     s3Object.setDataFile(dataFile);
 
@@ -756,7 +757,7 @@ public class FileStore {
    * @param uploadId id of the upload
    * @param partNumber number of the part to store
    * @param inputStream file data to be stored
-   * @param useV4Signing If {@code true}, V4-style signing is enabled.
+   * @param useV4ChunkedWithSigningFormat If {@code true}, V4-style signing is enabled.
    *
    * @return the md5 hash of this part
    *
@@ -767,9 +768,9 @@ public class FileStore {
       final String uploadId,
       final String partNumber,
       final InputStream inputStream,
-      final boolean useV4Signing) throws IOException {
+      final boolean useV4ChunkedWithSigningFormat) throws IOException {
     try (final DigestInputStream digestingInputStream =
-        new DigestInputStream(wrapStream(inputStream, useV4Signing),
+        new DigestInputStream(wrapStream(inputStream, useV4ChunkedWithSigningFormat),
             MessageDigest.getInstance("MD5"))) {
       inputStreamToFile(digestingInputStream,
           Paths.get(rootFolder.getAbsolutePath(), bucketName, fileName, uploadId,
@@ -867,7 +868,7 @@ public class FileStore {
     });
   }
 
-  private String[] listAndSortPartsInFromDirectory(File partFolder) {
+  private String[] listAndSortPartsInFromDirectory(final File partFolder) {
     final String[] partNames = partFolder.list((dir, name) -> name.endsWith(PART_SUFFIX));
 
     Arrays.sort(partNames,
@@ -922,10 +923,10 @@ public class FileStore {
     return Paths.get(rootFolder.getAbsolutePath(), bucketName, fileName, uploadId).toFile();
   }
 
-  private List<Part> arrangeSeparateParts(File[] files, final String bucketName,
+  private List<Part> arrangeSeparateParts(final File[] files, final String bucketName,
       final String fileName, final String uploadId) {
 
-    List<Part> parts = new ArrayList<>();
+    final List<Part> parts = new ArrayList<>();
 
     for (int i = 0; i < files.length; i++) {
       final String filePartPath = concatUploadIdAndPartFileName(files[i], uploadId);
@@ -953,13 +954,13 @@ public class FileStore {
     try (final InputStream is = FileUtils.openInputStream(currentFilePart)) {
       final String partMd5 = DigestUtils.md5Hex(is);
       return String.format("%s-%s", partMd5, partNumber);
-    } catch (IOException e) {
+    } catch (final IOException e) {
       LOG.error("Hash could not be calculated. File access did not succeed", e);
       return "";
     }
   }
 
-  private String concatUploadIdAndPartFileName(File file, String uploadId) {
+  private String concatUploadIdAndPartFileName(final File file, final String uploadId) {
     return String.format("%s/%s", uploadId, file.getName());
   }
 

--- a/server/src/test/java/com/adobe/testing/s3mock/its/AmazonClientUploadIT.java
+++ b/server/src/test/java/com/adobe/testing/s3mock/its/AmazonClientUploadIT.java
@@ -87,7 +87,6 @@ import com.amazonaws.services.s3.transfer.model.UploadResult;
 import java.io.ByteArrayInputStream;
 import java.io.File;
 import java.io.FileInputStream;
-import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.URL;
@@ -106,18 +105,19 @@ import javax.net.ssl.SSLContext;
 import javax.net.ssl.TrustManager;
 import javax.net.ssl.X509TrustManager;
 import org.apache.commons.codec.digest.DigestUtils;
-import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
 
 /**
  * Test the application using the AmazonS3 client.
  */
-public class AmazonClientUploadIT extends S3TestBase {
+class AmazonClientUploadIT extends S3TestBase {
   /**
    * Verify that buckets can be created and listed.
    */
   @Test
-  public void shouldCreateBucketAndListAllBuckets() {
+  void shouldCreateBucketAndListAllBuckets() {
     // the returned creation date might strip off the millisecond-part, resulting in rounding down
     // and account for a clock-skew in the Docker container of up to a minute.
     final Date creationDate = new Date((System.currentTimeMillis() / 1000) * 1000 - 60000);
@@ -144,7 +144,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Verifies that default Buckets got created after S3 Mock was bootstrapped.
    */
   @Test
-  public void defaultBucketsGotCreated() {
+  void defaultBucketsGotCreated() {
     final List<Bucket> buckets = s3Client.listBuckets();
     final Set<String> bucketNames = buckets.stream().map(Bucket::getName)
         .filter(INITIAL_BUCKET_NAMES::contains).collect(Collectors.toSet());
@@ -158,7 +158,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Verifies {@link AmazonS3#doesObjectExist}.
    */
   @Test
-  public void putObjectWhereKeyContainsPathFragments() {
+  void putObjectWhereKeyContainsPathFragments() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -174,7 +174,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception if FileStreams can not be read
    */
   @Test
-  public void shouldUploadAndListV2Objects() throws Exception {
+  void shouldUploadAndListV2Objects() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -193,13 +193,7 @@ public class AmazonClientUploadIT extends S3TestBase {
     for (final S3ObjectSummary objectSummary : listResult.getObjectSummaries()) {
       assertThat(objectSummary.getKey(), containsString(uploadFile.getName()));
       final S3Object s3Object = s3Client.getObject(BUCKET_NAME, objectSummary.getKey());
-      final InputStream uploadFileIs = new FileInputStream(uploadFile);
-      final String uploadHash = HashUtil.getDigest(uploadFileIs);
-      final String downloadedHash = HashUtil.getDigest(s3Object.getObjectContent());
-      uploadFileIs.close();
-      s3Object.close();
-      assertThat("Up- and downloaded Files should have equal Hashes", uploadHash,
-          is(equalTo(downloadedHash)));
+      verifyObjectContent(uploadFile, s3Object);
     }
   }
 
@@ -208,12 +202,22 @@ public class AmazonClientUploadIT extends S3TestBase {
    *
    * @throws Exception if FileStreams can not be read
    */
-  @Test
-  public void shouldUploadAndDownloadObject() throws Exception {
+  @ParameterizedTest(
+      name = ParameterizedTest.INDEX_PLACEHOLDER + " uploadWithSigning={0}, uploadChunked={1}")
+  @CsvSource(value = {"true, true", "true, false", "false, true", "false, false"})
+  void shouldUploadAndDownloadObject(final boolean uploadWithSigning,
+      final boolean uploadChunked)
+      throws Exception {
+    s3Client.createBucket(BUCKET_NAME);
+
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
-    s3Client.createBucket(BUCKET_NAME);
-    s3Client.putObject(new PutObjectRequest(BUCKET_NAME, uploadFile.getName(), uploadFile));
+    final AmazonS3 uploadClient = defaultTestAmazonS3ClientBuilder()
+        .withPayloadSigningEnabled(uploadWithSigning)
+        .withChunkedEncodingDisabled(uploadChunked)
+        .build();
+
+    uploadClient.putObject(new PutObjectRequest(BUCKET_NAME, uploadFile.getName(), uploadFile));
 
     final S3Object s3Object = s3Client.getObject(BUCKET_NAME, uploadFile.getName());
 
@@ -223,11 +227,10 @@ public class AmazonClientUploadIT extends S3TestBase {
   /**
    * Uses weird, but valid characters in the key used to store an object.
    *
-   * @see #shouldUploadAndDownloadObject()
    * @throws Exception if FileStreams can not be read
    */
   @Test
-  public void shouldTolerateWeirdCharactersInObjectKey() throws Exception {
+  void shouldTolerateWeirdCharactersInObjectKey() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -244,7 +247,7 @@ public class AmazonClientUploadIT extends S3TestBase {
   }
 
   private void verifyObjectContent(final File uploadFile, final S3Object s3Object)
-      throws FileNotFoundException, NoSuchAlgorithmException, IOException {
+      throws NoSuchAlgorithmException, IOException {
     final InputStream uploadFileIs = new FileInputStream(uploadFile);
     final String uploadHash = HashUtil.getDigest(uploadFileIs);
     final String downloadedHash = HashUtil.getDigest(s3Object.getObjectContent());
@@ -258,11 +261,9 @@ public class AmazonClientUploadIT extends S3TestBase {
   /**
    * Uses weird, but valid characters in the key used to store an object. Verifies
    * that ListObject returns the correct object names.
-   *
-   * @throws Exception if FileStreams can not be read
    */
   @Test
-  public void shouldListWithCorrectObjectNames() throws Exception {
+  void shouldListWithCorrectObjectNames() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -283,11 +284,9 @@ public class AmazonClientUploadIT extends S3TestBase {
 
   /**
    * Same as {@link #shouldListWithCorrectObjectNames()} but for V2 API.
-   *
-   * @throws Exception if FileStreams can not be read
    */
   @Test
-  public void shouldListV2WithCorrectObjectNames() throws Exception {
+  void shouldListV2WithCorrectObjectNames() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -319,11 +318,9 @@ public class AmazonClientUploadIT extends S3TestBase {
    *
    * <p>This isn't the greatest way to test this functionality, however, there
    * is currently no low-level testing infrastructure in place.
-   *
-   * @throws Exception if FileStreams can not be read
    */
   @Test
-  public void shouldHonorEncodingType() throws Exception {
+  void shouldHonorEncodingType() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -338,21 +335,16 @@ public class AmazonClientUploadIT extends S3TestBase {
 
     // we expect an SdkClientException wich a message pointing to XML
     // parsing issues.
-    assertThat(
-        assertThrows(SdkClientException.class, () -> {
-          s3Client.listObjects(lor);
-        }).getMessage(),
+    assertThat(assertThrows(SdkClientException.class, () -> s3Client.listObjects(lor)).getMessage(),
         containsString("Failed to parse XML document")
     );
   }
 
   /**
    * The same as {@link #shouldHonorEncodingType()} but for V2 API.
-   *
-   * @throws Exception if FileStreams can not be read
    */
   @Test
-  public void shouldHonorEncodingTypeV2() throws Exception {
+  void shouldHonorEncodingTypeV2() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -370,9 +362,7 @@ public class AmazonClientUploadIT extends S3TestBase {
     // we expect an SdkClientException wich a message pointing to XML
     // parsing issues.
     assertThat(
-        assertThrows(SdkClientException.class, () -> {
-          s3Client.listObjectsV2(lorv2);
-        }).getMessage(),
+        assertThrows(SdkClientException.class, () -> s3Client.listObjectsV2(lorv2)).getMessage(),
         containsString("Failed to parse XML document")
     );
   }
@@ -383,7 +373,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception if FileStreams can not be read
    */
   @Test
-  public void shouldUploadAndDownloadStream() throws Exception {
+  void shouldUploadAndDownloadStream() throws Exception {
     s3Client.createBucket(BUCKET_NAME);
     final String resourceId = randomUUID().toString();
     final String contentEncoding = "gzip";
@@ -421,7 +411,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests if Object can be uploaded with KMS and Metadata can be retrieved.
    */
   @Test
-  public void shouldUploadWithEncryption() {
+  void shouldUploadWithEncryption() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     final String objectKey = UPLOAD_FILE_NAME;
     s3Client.createBucket(BUCKET_NAME);
@@ -449,18 +439,16 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests if Object can be uploaded with wrong KMS Key.
    */
   @Test
-  public void shouldNotUploadWithWrongEncryptionKey() {
+  void shouldNotUploadWithWrongEncryptionKey() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     s3Client.createBucket(BUCKET_NAME);
     final PutObjectRequest putObjectRequest =
         new PutObjectRequest(BUCKET_NAME, UPLOAD_FILE_NAME, uploadFile);
     putObjectRequest.setSSEAwsKeyManagementParams(new SSEAwsKeyManagementParams(TEST_WRONG_KEYREF));
 
-    final AmazonS3Exception e = Assertions.assertThrows(AmazonS3Exception.class, () -> {
-      s3Client.putObject(putObjectRequest);
-    });
-
-    assertThat(e.getMessage(),
+    assertThat(
+        assertThrows(AmazonS3Exception.class, () -> s3Client.putObject(putObjectRequest))
+            .getMessage(),
         containsString("Status Code: 400; Error Code: KMS.NotFoundException"));
   }
 
@@ -468,7 +456,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests if Object can be uploaded with wrong KMS Key.
    */
   @Test
-  public void shouldNotUploadStreamingWithWrongEncryptionKey() {
+  void shouldNotUploadStreamingWithWrongEncryptionKey() {
     final byte[] bytes = UPLOAD_FILE_NAME.getBytes();
     final InputStream stream = new ByteArrayInputStream(bytes);
     final String objectKey = randomUUID().toString();
@@ -479,11 +467,9 @@ public class AmazonClientUploadIT extends S3TestBase {
         new PutObjectRequest(BUCKET_NAME, objectKey, stream, metadata);
     putObjectRequest.setSSEAwsKeyManagementParams(new SSEAwsKeyManagementParams(TEST_WRONG_KEYREF));
 
-    final AmazonS3Exception e = Assertions.assertThrows(AmazonS3Exception.class, () -> {
-      s3Client.putObject(putObjectRequest);
-    });
-
-    assertThat(e.getMessage(),
+    assertThat(
+        assertThrows(AmazonS3Exception.class, () -> s3Client.putObject(putObjectRequest))
+            .getMessage(),
         containsString("Status Code: 400; Error Code: KMS.NotFoundException"));
   }
 
@@ -494,7 +480,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception if an Exception occurs
    */
   @Test
-  public void shouldCopyObject() throws Exception {
+  void shouldCopyObject() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     final String sourceKey = UPLOAD_FILE_NAME;
     final String destinationBucketName = "destinationbucket";
@@ -526,7 +512,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception if an Exception occurs
    */
   @Test
-  public void shouldCopyObjectToKeyNeedingEscaping() throws Exception {
+  void shouldCopyObjectToKeyNeedingEscaping() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     final String sourceKey = UPLOAD_FILE_NAME;
     final String destinationBucketName = "destinationbucket";
@@ -558,7 +544,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception if an Exception occurs
    */
   @Test
-  public void shouldCopyObjectFromKeyNeedingEscaping() throws Exception {
+  void shouldCopyObjectFromKeyNeedingEscaping() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     final String sourceKey = "some escape-worthy characters %$@ " + UPLOAD_FILE_NAME;
     final String destinationBucketName = "destinationbucket";
@@ -590,7 +576,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception if an Exception occurs
    */
   @Test
-  public void shouldCopyObjectEncrypted() throws Exception {
+  void shouldCopyObjectEncrypted() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     final String sourceKey = UPLOAD_FILE_NAME;
     final String destinationBucketName = "destinationbucket";
@@ -620,7 +606,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests that an object wont be copied with wrong encryption Key.
    */
   @Test
-  public void shouldNotObjectCopyWithWrongEncryptionKey() {
+  void shouldNotObjectCopyWithWrongEncryptionKey() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     final String sourceKey = UPLOAD_FILE_NAME;
     final String destinationBucketName = "destinationbucket";
@@ -635,11 +621,8 @@ public class AmazonClientUploadIT extends S3TestBase {
     copyObjectRequest
         .setSSEAwsKeyManagementParams(new SSEAwsKeyManagementParams(TEST_WRONG_KEYREF));
 
-    final AmazonS3Exception e = Assertions.assertThrows(AmazonS3Exception.class, () -> {
-      s3Client.copyObject(copyObjectRequest);
-    });
-
-    assertThat(e.getMessage(),
+    assertThat(assertThrows(AmazonS3Exception.class, () -> s3Client.copyObject(copyObjectRequest))
+            .getMessage(),
         containsString("Status Code: 400; Error Code: KMS.NotFoundException"));
   }
 
@@ -647,7 +630,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Creates a bucket and checks if it exists using {@link AmazonS3Client#doesBucketExist(String)}.
    */
   @Test
-  public void bucketShouldExist() {
+  void bucketShouldExist() {
     s3Client.createBucket(BUCKET_NAME);
 
     final Boolean doesBucketExist = s3Client.doesBucketExistV2(BUCKET_NAME);
@@ -661,7 +644,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Checks if {@link AmazonS3Client#doesBucketExistV2(String)} is false on a not existing Bucket.
    */
   @Test
-  public void bucketShouldNotExist() {
+  void bucketShouldNotExist() {
     final Boolean doesBucketExist = s3Client.doesBucketExistV2(BUCKET_NAME);
 
     assertThat(String.format("The bucket, '%s', should not exist!", BUCKET_NAME), doesBucketExist,
@@ -672,7 +655,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests if the Metadata of an existing file can be retrieved.
    */
   @Test
-  public void shouldGetObjectMetadata() {
+  void shouldGetObjectMetadata() {
     final String nonExistingFileName = "nonExistingFileName";
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     s3Client.createBucket(BUCKET_NAME);
@@ -694,36 +677,34 @@ public class AmazonClientUploadIT extends S3TestBase {
     assertThat("User metadata should be identical!", metadataExisting.getUserMetadata(),
         is(equalTo(objectMetadata.getUserMetadata())));
 
-    final AmazonS3Exception e = Assertions.assertThrows(AmazonS3Exception.class, () -> {
-      s3Client.getObjectMetadata(BUCKET_NAME, nonExistingFileName);
-    });
-
-    assertThat(e.getMessage(), containsString("Status Code: 404"));
+    assertThat(
+        assertThrows(AmazonS3Exception.class,
+            () -> s3Client.getObjectMetadata(BUCKET_NAME, nonExistingFileName)).getMessage(),
+        containsString("Status Code: 404"));
   }
 
   /**
    * Tests if an object can be deleted.
    */
   @Test
-  public void shouldDeleteObject() {
+  void shouldDeleteObject() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     s3Client.createBucket(BUCKET_NAME);
 
     s3Client.putObject(new PutObjectRequest(BUCKET_NAME, UPLOAD_FILE_NAME, uploadFile));
     s3Client.deleteObject(BUCKET_NAME, UPLOAD_FILE_NAME);
 
-    final AmazonS3Exception e = Assertions.assertThrows(AmazonS3Exception.class, () -> {
-      s3Client.getObject(BUCKET_NAME, UPLOAD_FILE_NAME);
-    });
-
-    assertThat(e.getMessage(), containsString("Status Code: 404"));
+    assertThat(
+        assertThrows(AmazonS3Exception.class,
+            () -> s3Client.getObject(BUCKET_NAME, UPLOAD_FILE_NAME)).getMessage(),
+        containsString("Status Code: 404"));
   }
 
   /**
    * Tests if an object can be deleted.
    */
   @Test
-  public void shouldBatchDeleteObjects() {
+  void shouldBatchDeleteObjects() {
     final File uploadFile1 = new File(UPLOAD_FILE_NAME);
     final File uploadFile2 = new File(UPLOAD_FILE_NAME);
     final File uploadFile3 = new File(UPLOAD_FILE_NAME);
@@ -756,18 +737,16 @@ public class AmazonClientUploadIT extends S3TestBase {
         delObjRes.getDeletedObjects().stream().map(DeletedObject::getKey).collect(toList()),
         contains(file1, file2, file3));
 
-    final AmazonS3Exception e = Assertions.assertThrows(AmazonS3Exception.class, () -> {
-      s3Client.getObject(BUCKET_NAME, UPLOAD_FILE_NAME);
-    });
-
-    assertThat(e.getMessage(), containsString("Status Code: 404"));
+    assertThat(assertThrows(AmazonS3Exception.class,
+        () -> s3Client.getObject(BUCKET_NAME, UPLOAD_FILE_NAME)).getMessage(),
+        containsString("Status Code: 404"));
   }
 
   /**
    * Tests that a bucket can be deleted.
    */
   @Test
-  public void shouldDeleteBucket() {
+  void shouldDeleteBucket() {
     s3Client.createBucket(BUCKET_NAME);
     s3Client.deleteBucket(BUCKET_NAME);
 
@@ -781,7 +760,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * <p>For more detailed tests of the List Objects API see {@link ListObjectIT}.
    */
   @Test
-  public void shouldGetObjectListing() {
+  void shouldGetObjectListing() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
     s3Client.createBucket(BUCKET_NAME);
     s3Client.putObject(new PutObjectRequest(BUCKET_NAME, UPLOAD_FILE_NAME, uploadFile));
@@ -803,7 +782,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception not expected.
    */
   @Test
-  public void shouldUploadInParallel() throws Exception {
+  void shouldUploadInParallel() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -823,7 +802,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests if user metadata can be passed by multipart upload.
    */
   @Test
-  public void shouldPassUserMetadataWithMultipartUploads() {
+  void shouldPassUserMetadataWithMultipartUploads() {
     s3Client.createBucket(BUCKET_NAME);
 
     final File uploadFile = new File(UPLOAD_FILE_NAME);
@@ -862,7 +841,7 @@ public class AmazonClientUploadIT extends S3TestBase {
   }
 
   @Test
-  public void shouldInitiateMultipartAndRetrieveParts() throws IOException {
+  void shouldInitiateMultipartAndRetrieveParts() throws IOException {
     s3Client.createBucket(BUCKET_NAME);
 
     final File uploadFile = new File(UPLOAD_FILE_NAME);
@@ -877,7 +856,7 @@ public class AmazonClientUploadIT extends S3TestBase {
     final String uploadId = initiateMultipartUploadResult.getUploadId();
     final String key = initiateMultipartUploadResult.getKey();
 
-    final UploadPartResult uploadPartResult = s3Client.uploadPart(new UploadPartRequest()
+    s3Client.uploadPart(new UploadPartRequest()
         .withBucketName(initiateMultipartUploadResult.getBucketName())
         .withKey(initiateMultipartUploadResult.getKey())
         .withUploadId(uploadId)
@@ -887,7 +866,7 @@ public class AmazonClientUploadIT extends S3TestBase {
         .withPartSize(uploadFile.length())
         .withLastPart(true));
 
-    ListPartsRequest listPartsRequest = new ListPartsRequest(
+    final ListPartsRequest listPartsRequest = new ListPartsRequest(
         BUCKET_NAME,
         key,
         uploadId
@@ -895,7 +874,7 @@ public class AmazonClientUploadIT extends S3TestBase {
     final PartListing partListing = s3Client.listParts(listPartsRequest);
 
     assertThat("Part listing should be 1", partListing.getParts().size(), is(1));
-    PartSummary partSummary = partListing.getParts().get(0);
+    final PartSummary partSummary = partListing.getParts().get(0);
 
     assertThat("Etag should match", partSummary.getETag(), is(expectedEtag));
     assertThat("Part number should match", partSummary.getPartNumber(), is(1));
@@ -907,7 +886,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests if not yet completed / aborted multipart uploads are listed.
    */
   @Test
-  public void shouldListMultipartUploads() {
+  void shouldListMultipartUploads() {
     s3Client.createBucket(BUCKET_NAME);
 
     assertThat(s3Client.listMultipartUploads(new ListMultipartUploadsRequest(BUCKET_NAME))
@@ -931,7 +910,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Tests if a multipart upload can be aborted.
    */
   @Test
-  public void shouldAbortMultipartUpload() {
+  void shouldAbortMultipartUpload() {
     s3Client.createBucket(BUCKET_NAME);
 
     assertThat(s3Client.listMultipartUploads(new ListMultipartUploadsRequest(BUCKET_NAME))
@@ -957,7 +936,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * @throws Exception not expected
    */
   @Test
-  public void checkRangeDownloads() throws Exception {
+  void checkRangeDownloads() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -986,7 +965,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Verifies multipart copy.
    */
   @Test
-  public void multipartCopy() throws InterruptedException {
+  void multipartCopy() throws InterruptedException {
     final int contentLen = 3 * _1MB;
 
     final ObjectMetadata objectMetadata = new ObjectMetadata();
@@ -1024,7 +1003,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Creates a bucket, stores a file, adds tags, retrieves tags and checks them for consistency.
    */
   @Test
-  public void shouldAddAndRetrieveTags() {
+  void shouldAddAndRetrieveTags() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -1061,7 +1040,7 @@ public class AmazonClientUploadIT extends S3TestBase {
    * Creates a bucket, stores a file with tags, retrieves tags and checks them for consistency.
    */
   @Test
-  public void canAddTagsOnPutObject() {
+  void canAddTagsOnPutObject() {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -1070,31 +1049,31 @@ public class AmazonClientUploadIT extends S3TestBase {
     tagList.add(new Tag("foo", "bar"));
 
     final PutObjectRequest putObjectRequest =
-            new PutObjectRequest(BUCKET_NAME, uploadFile.getName(), uploadFile)
-                    .withTagging(new ObjectTagging(tagList));
+        new PutObjectRequest(BUCKET_NAME, uploadFile.getName(), uploadFile)
+            .withTagging(new ObjectTagging(tagList));
 
     s3Client.putObject(putObjectRequest);
 
     final S3Object s3Object = s3Client.getObject(BUCKET_NAME, uploadFile.getName());
 
-    GetObjectTaggingRequest getObjectTaggingRequest = new GetObjectTaggingRequest(BUCKET_NAME,
-            s3Object.getKey());
+    final GetObjectTaggingRequest getObjectTaggingRequest = new GetObjectTaggingRequest(BUCKET_NAME,
+        s3Object.getKey());
 
-    GetObjectTaggingResult getObjectTaggingResult = s3Client
-            .getObjectTagging(getObjectTaggingRequest);
+    final GetObjectTaggingResult getObjectTaggingResult = s3Client
+        .getObjectTagging(getObjectTaggingRequest);
 
     // There should be 'foo:bar' here
     assertThat("Couldn't find that the tag that was placed",
-            getObjectTaggingResult.getTagSet().size(), is(1));
+        getObjectTaggingResult.getTagSet().size(), is(1));
     assertThat("The vaule of the tag placed did not match",
-            getObjectTaggingResult.getTagSet().get(0).getValue(), is("bar"));
+        getObjectTaggingResult.getTagSet().get(0).getValue(), is("bar"));
   }
 
   /**
    * Creates a bucket, stores a file, get files with eTag requrements.
    */
   @Test
-  public void shouldCreateAndRespectEtag() throws Exception {
+  void shouldCreateAndRespectEtag() throws Exception {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
     s3Client.createBucket(BUCKET_NAME);
@@ -1148,7 +1127,7 @@ public class AmazonClientUploadIT extends S3TestBase {
   }
 
   @Test
-  public void generatePresignedUrlWithResponseHeaderOverrides()
+  void generatePresignedUrlWithResponseHeaderOverrides()
       throws IOException, NoSuchAlgorithmException, KeyManagementException {
     final File uploadFile = new File(UPLOAD_FILE_NAME);
 
@@ -1188,18 +1167,21 @@ public class AmazonClientUploadIT extends S3TestBase {
     urlConnection.getInputStream().close();
   }
 
-  private URLConnection openUrlConnection(URL resourceUrl)
+  private URLConnection openUrlConnection(final URL resourceUrl)
       throws NoSuchAlgorithmException, KeyManagementException, IOException {
     final TrustManager[] trustAllCerts = new TrustManager[] {
         new X509TrustManager() {
+          @Override
           public X509Certificate[] getAcceptedIssuers() {
             return null;
           }
 
-          public void checkClientTrusted(X509Certificate[] certs, String authType) {
+          @Override
+          public void checkClientTrusted(final X509Certificate[] certs, final String authType) {
           }
 
-          public void checkServerTrusted(X509Certificate[] certs, String authType) {
+          @Override
+          public void checkServerTrusted(final X509Certificate[] certs, final String authType) {
           }
         }
     };

--- a/server/src/test/java/com/adobe/testing/s3mock/its/S3TestBase.java
+++ b/server/src/test/java/com/adobe/testing/s3mock/its/S3TestBase.java
@@ -78,16 +78,17 @@ abstract class S3TestBase {
    */
   @BeforeEach
   public void prepareS3Client() {
-    final BasicAWSCredentials credentials = new BasicAWSCredentials("foo", "bar");
+    s3Client = defaultTestAmazonS3ClientBuilder().build();
+  }
 
-    s3Client = AmazonS3ClientBuilder.standard()
-        .withCredentials(new AWSStaticCredentialsProvider(credentials))
+  protected AmazonS3ClientBuilder defaultTestAmazonS3ClientBuilder() {
+    return AmazonS3ClientBuilder.standard()
+        .withCredentials(new AWSStaticCredentialsProvider(new BasicAWSCredentials("foo", "bar")))
         .withClientConfiguration(ignoringInvalidSslCertificates(new ClientConfiguration()))
         .withEndpointConfiguration(
             new AwsClientBuilder.EndpointConfiguration("https://" + getHost() + ":" + getPort(),
                 "us-east-1"))
-        .enablePathStyleAccess()
-        .build();
+        .enablePathStyleAccess();
   }
 
   /**

